From d6798c05437d91387d26de8171b910fd6beac975 Mon Sep 17 00:00:00 2001
From: Lu Jialin <lujialin4@huawei.com>
Date: Wed, 17 May 2023 12:33:51 +0800
Subject: [PATCH 020/263] [Huawei] sched/psi: Fix illegal memory access in
 psi_account_irqtime()

Offering: HULK
HULK inclusion
category: feature
bugzilla: 188780

-------------------------------

When psi_disabled is true, cgroup->psi.pcpu wouldn't be initialized. Under
the circumstance, a NULL pointer dereference may happen in
psi_account_irqtime(). So just return in psi_account_irqtime(), when
psi_disabled is true.

Signed-off-by: Lu Jialin <lujialin4@huawei.com>
---
 kernel/sched/psi.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/kernel/sched/psi.c b/kernel/sched/psi.c
index 0485f5d06382..d96113d7ea52 100644
--- a/kernel/sched/psi.c
+++ b/kernel/sched/psi.c
@@ -936,6 +936,9 @@ void psi_account_irqtime(struct task_struct *task, u32 delta)
 	struct psi_group_cpu *groupc;
 	u64 now;
 
+	if (static_branch_likely(&psi_disabled))
+		return;
+
 	if (!task->pid)
 		return;
 
-- 
2.17.1

